toolchain/check/impl_validation.cpp

```diff
@@ -0,0 +1,456 @@
+                      "the root self type nor the interface definition");
+    context.emitter().Emit(impl.latest_decl_id, FinalImplInvalidFile);
+    return true;
+  }
+
+  return false;
+}
+
+// The type structure each non-final `impl` must differ from all other non-final
+// `impl` for the same interface visible from the file.
+//
+// Returns true if an error was diagnosed.
+static auto DiagnoseNonFinalImplsWithSameTypeStructure(Context& context,
+                                                       const ImplInfo& impl_a,
+                                                       const ImplInfo& impl_b)
+    -> bool {
+  if (impl_a.type_structure == impl_b.type_structure) {
+    CARBON_DIAGNOSTIC(ImplNonFinalSameTypeStructure, Error,
+                      "found non-final `impl` with the same type "
+                      "structure as another non-final `impl`");
+    auto builder = context.emitter().Build(impl_b.latest_decl_id,
+                                           ImplNonFinalSameTypeStructure);
+    CARBON_DIAGNOSTIC(ImplNonFinalSameTypeStructureNote, Note,
+                      "other `impl` here");
+    builder.Note(impl_a.latest_decl_id, ImplNonFinalSameTypeStructureNote);
+    builder.Emit();
+    return true;
+  }
+
+  return false;
+}
+
+// An impl's self type and constraint can not match (as a lookup query)
+// against any final impl, or it would never be used instead of that
+// final impl.
+//
+// Returns true if an error was diagnosed.
+static auto DiagnoseUnmatchableNonFinalImplWithFinalImpl(
+    Context& context, SemIR::ImplId impl_a_id, const ImplInfo& impl_a,
+    SemIR::ImplId impl_b_id, const ImplInfo& impl_b) -> bool {
+  auto diagnose_unmatchable_impl = [&](const ImplInfo& query_impl,
+                                       SemIR::ImplId final_impl_id,
+                                       const ImplInfo& final_impl) -> bool {
+    if (LookupMatchesImpl(context, SemIR::LocId(query_impl.latest_decl_id),
+                          context.constant_values().Get(query_impl.self_id),
+                          query_impl.interface, final_impl_id)) {
+      CARBON_DIAGNOSTIC(ImplFinalOverlapsNonFinal, Error,
+                        "`impl` will never be used");
+      auto builder = context.emitter().Build(query_impl.latest_decl_id,
+                                             ImplFinalOverlapsNonFinal);
+      CARBON_DIAGNOSTIC(
+          ImplFinalOverlapsNonFinalNote, Note,
+          "`final impl` declared here would always be used instead");
+      builder.Note(final_impl.latest_decl_id, ImplFinalOverlapsNonFinalNote);
+      builder.Emit();
+      return true;
+    }
+    return false;
+  };
+
+  bool did_diagnose = false;
+  if (impl_a.is_final) {
+    did_diagnose = diagnose_unmatchable_impl(impl_b, impl_a_id, impl_a);
+  }
+  if (impl_b.is_final && !did_diagnose) {
+    did_diagnose = diagnose_unmatchable_impl(impl_a, impl_b_id, impl_b);
+  }
+  if (did_diagnose) {
+    return true;
+  }
+
+  return false;
```

Comment:
I feel like this code is more complicated than necessary if we know that exactly one of `impl_a.is_final` and `impl_b.is_final` is true:
```suggestion
  bool did_diagnose = false;
  if (impl_a.is_final) {
    CARBON_CHECK(!impl_b.is_final);
    retrun diagnose_unmatchable_impl(impl_b, impl_a_id, impl_a);
  } else {
    CARBON_CHECK(impl_b.is_final);
    return diagnose_unmatchable_impl(impl_a, impl_b_id, impl_b);
  }
```

Reply:
Yes this can be nicer now that I split it into a separate function, thanks.